In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np

#### Импортируем таблицу

In [2]:
df = pd.read_excel('dz-z2.xls')
df

card_nm  card_id receive_dt     use_dt
0       Beta  7637320 2018-11-02        NaT
1      Alpha  7635974 2018-11-02        NaT
2      Alpha  7635511 2018-11-02        NaT
3      Gamma  7635258 2018-11-02 2018-12-13
4       Beta  7634217 2018-11-02        NaT
5      Gamma  7633769 2018-11-02        NaT
6       Beta  7629486 2018-11-02        NaT
7      Alpha  7629329 2018-11-02        NaT
8      Alpha  7627083 2018-11-02        NaT
9       Beta  7620367 2018-11-02        NaT
10     Alpha  7618161 2018-11-02        NaT
11      Beta  7617613 2018-11-02        NaT
12      Beta  7617456 2018-11-02        NaT
13     Gamma  7617067 2018-11-02        NaT
14     Gamma  7616202 2018-11-02        NaT
15     Alpha  7615628 2018-11-02 2018-11-11
16     Gamma  7613189 2018-11-02        NaT
17     Alpha  7612010 2018-11-02        NaT
18     Alpha  7611535 2018-11-02        NaT
19     Alpha  7611274 2018-11-02        NaT
20     Alpha  7609628 2018-11-02        NaT
21     Alpha  7608779 2018-11-02        NaT
22      Beta  7607799 2018-11-02        NaT
23     Alpha  7607682 2018-11-02        NaT
24      Beta  7606990 2018-11-02        NaT
25     Alpha  7599622 2018-11-02        NaT
26     Alpha  7598493 2018-11-03        NaT
27      Beta  7598421 2018-11-03        NaT
28     Gamma  7597756 2018-11-02        NaT
29     Alpha  7597753 2018-11-03 2018-11-05
...      ...      ...        ...        ...
6734   Alpha   671363 2019-04-08        NaT
6735   Alpha   670314 2019-04-08        NaT
6736   Gamma   670203 2019-04-07        NaT
6737   Alpha   669418 2019-04-07        NaT
6738   Alpha   668863 2019-04-08        NaT
6739   Alpha   668573 2019-04-08        NaT
6740   Alpha   668300 2019-04-07        NaT
6741   Alpha   668018 2019-04-08        NaT
6742    Beta   667984 2019-04-08        NaT
6743   Gamma   667484 2019-04-07        NaT
6744    Beta   667338 2019-04-08        NaT
6745   Alpha   666341 2019-04-08        NaT
6746   Alpha   666123 2019-04-08        NaT
6747   Alpha   665817 2019-04-08        NaT
6748   Alpha   665510 2019-04-08        NaT
6749   Alpha   664497 2019-04-07        NaT
6750   Gamma   663972 2019-04-08        NaT
6751   Gamma   663741 2019-04-07        NaT
6752   Alpha   663417 2019-04-07        NaT
6753   Gamma   663026 2019-04-08        NaT
6754   Alpha   655878 2019-04-08        NaT
6755   Alpha   654443 2019-04-08        NaT
6756   Alpha   647634 2019-04-08        NaT
6757   Alpha   645987 2019-04-08        NaT
6758   Alpha   638571 2019-04-08        NaT
6759   Alpha   637512 2019-04-08        NaT
6760    Beta   637397 2019-04-08        NaT
6761   Alpha   628982 2019-04-08 2019-04-10
6762    Beta   625773 2019-04-08        NaT
6763   Alpha   622700 2019-04-08        NaT

[6764 rows x 4 columns]

#### use_dt для анализа не очень подходит, добавим столбец флагов использования сертификата

In [3]:
df['use'] = ((pd.isnull(df.use_dt))==False)*1
df

card_nm  card_id receive_dt     use_dt  use
0       Beta  7637320 2018-11-02        NaT    0
1      Alpha  7635974 2018-11-02        NaT    0
2      Alpha  7635511 2018-11-02        NaT    0
3      Gamma  7635258 2018-11-02 2018-12-13    1
4       Beta  7634217 2018-11-02        NaT    0
5      Gamma  7633769 2018-11-02        NaT    0
6       Beta  7629486 2018-11-02        NaT    0
7      Alpha  7629329 2018-11-02        NaT    0
8      Alpha  7627083 2018-11-02        NaT    0
9       Beta  7620367 2018-11-02        NaT    0
10     Alpha  7618161 2018-11-02        NaT    0
11      Beta  7617613 2018-11-02        NaT    0
12      Beta  7617456 2018-11-02        NaT    0
13     Gamma  7617067 2018-11-02        NaT    0
14     Gamma  7616202 2018-11-02        NaT    0
15     Alpha  7615628 2018-11-02 2018-11-11    1
16     Gamma  7613189 2018-11-02        NaT    0
17     Alpha  7612010 2018-11-02        NaT    0
18     Alpha  7611535 2018-11-02        NaT    0
19     Alpha  7611274 2018-11-02        NaT    0
20     Alpha  7609628 2018-11-02        NaT    0
21     Alpha  7608779 2018-11-02        NaT    0
22      Beta  7607799 2018-11-02        NaT    0
23     Alpha  7607682 2018-11-02        NaT    0
24      Beta  7606990 2018-11-02        NaT    0
25     Alpha  7599622 2018-11-02        NaT    0
26     Alpha  7598493 2018-11-03        NaT    0
27      Beta  7598421 2018-11-03        NaT    0
28     Gamma  7597756 2018-11-02        NaT    0
29     Alpha  7597753 2018-11-03 2018-11-05    1
...      ...      ...        ...        ...  ...
6734   Alpha   671363 2019-04-08        NaT    0
6735   Alpha   670314 2019-04-08        NaT    0
6736   Gamma   670203 2019-04-07        NaT    0
6737   Alpha   669418 2019-04-07        NaT    0
6738   Alpha   668863 2019-04-08        NaT    0
6739   Alpha   668573 2019-04-08        NaT    0
6740   Alpha   668300 2019-04-07        NaT    0
6741   Alpha   668018 2019-04-08        NaT    0
6742    Beta   667984 2019-04-08        NaT    0
6743   Gamma   667484 2019-04-07        NaT    0
6744    Beta   667338 2019-04-08        NaT    0
6745   Alpha   666341 2019-04-08        NaT    0
6746   Alpha   666123 2019-04-08        NaT    0
6747   Alpha   665817 2019-04-08        NaT    0
6748   Alpha   665510 2019-04-08        NaT    0
6749   Alpha   664497 2019-04-07        NaT    0
6750   Gamma   663972 2019-04-08        NaT    0
6751   Gamma   663741 2019-04-07        NaT    0
6752   Alpha   663417 2019-04-07        NaT    0
6753   Gamma   663026 2019-04-08        NaT    0
6754   Alpha   655878 2019-04-08        NaT    0
6755   Alpha   654443 2019-04-08        NaT    0
6756   Alpha   647634 2019-04-08        NaT    0
6757   Alpha   645987 2019-04-08        NaT    0
6758   Alpha   638571 2019-04-08        NaT    0
6759   Alpha   637512 2019-04-08        NaT    0
6760    Beta   637397 2019-04-08        NaT    0
6761   Alpha   628982 2019-04-08 2019-04-10    1
6762    Beta   625773 2019-04-08        NaT    0
6763   Alpha   622700 2019-04-08        NaT    0

[6764 rows x 5 columns]

#### Посмотрим на конверсии по различным сертификатам:

In [5]:
df1 = df.groupby('card_nm')['use'].sum().to_frame()
df2 = df.groupby('card_nm')['use'].count().to_frame()
df_conv = pd.merge(df2, df1, left_on='card_nm', right_on='card_nm')
df_conv.rename(columns={'use_x': 'issued', 'use_y': 'used'}, inplace=True)
df_conv['issued_to_used'] = (df_conv['used']/df_conv['issued'])*100
df_conv = df_conv.round({'issued_to_used': 0})
df_conv

issued  used  issued_to_used
card_nm                              
Alpha      3570   812            23.0
Beta       1272    89             7.0
Gamma      1922   693            36.0

#### Видно, что лучше всего используются сертификаты Gamma -- 36% использования. Меньше всего используют сертификаты Beta -- 7% использования. Посмотрим использование сертификатов в динамике.

#### Создадим отдельные подтаблицы для каждого месяца. Для апреля не создаем, т.к. он еще не закончился.

In [6]:
df_nov = df[(df.receive_dt >= '2018-11-01') & (df.receive_dt < '2018-12-01')]
df_dec = df[(df.receive_dt >= '2018-12-01') & (df.receive_dt < '2019-01-01')]
df_jan = df[(df.receive_dt >= '2019-01-01') & (df.receive_dt < '2019-02-01')]
df_feb = df[(df.receive_dt >= '2019-02-01') & (df.receive_dt < '2019-03-01')]
df_mar = df[(df.receive_dt >= '2019-03-01') & (df.receive_dt < '2019-04-01')]

#### Ноябрь

In [7]:
df1 = df_nov.groupby('card_nm')['use'].sum().to_frame()
df2 = df_nov.groupby('card_nm')['use'].count().to_frame()
df_nov_conv = pd.merge(df2, df1, left_on='card_nm', right_on='card_nm')
df_nov_conv.rename(columns={'use_x': 'issued', 'use_y': 'used'}, inplace=True)
df_nov_conv['issued_to_used'] = (df_nov_conv['used']/df_nov_conv['issued'])*100
df_nov_conv = df_nov_conv.round({'issued_to_used': 0})
df_nov_conv

issued  used  issued_to_used
card_nm                              
Alpha       546   155            28.0
Beta        295    18             6.0
Gamma       342   157            46.0

#### Декабрь

In [8]:
df1 = df_dec.groupby('card_nm')['use'].sum().to_frame()
df2 = df_dec.groupby('card_nm')['use'].count().to_frame()
df_dec_conv = pd.merge(df2, df1, left_on='card_nm', right_on='card_nm')
df_dec_conv.rename(columns={'use_x': 'issued', 'use_y': 'used'}, inplace=True)
df_dec_conv['issued_to_used'] = (df_dec_conv['used']/df_dec_conv['issued'])*100
df_dec_conv = df_dec_conv.round({'issued_to_used': 0})
df_dec_conv

issued  used  issued_to_used
card_nm                              
Alpha       610   159            26.0
Beta        220    12             5.0
Gamma       364   171            47.0

#### Январь

In [9]:
df1 = df_jan.groupby('card_nm')['use'].sum().to_frame()
df2 = df_jan.groupby('card_nm')['use'].count().to_frame()
df_jan_conv = pd.merge(df2, df1, left_on='card_nm', right_on='card_nm')
df_jan_conv.rename(columns={'use_x': 'issued', 'use_y': 'used'}, inplace=True)
df_jan_conv['issued_to_used'] = (df_jan_conv['used']/df_jan_conv['issued'])*100
df_jan_conv = df_jan_conv.round({'issued_to_used': 0})
df_jan_conv

issued  used  issued_to_used
card_nm                              
Alpha       697   183            26.0
Beta        238    19             8.0
Gamma       424   197            46.0

#### Февраль

In [10]:
df1 = df_feb.groupby('card_nm')['use'].sum().to_frame()
df2 = df_feb.groupby('card_nm')['use'].count().to_frame()
df_feb_conv = pd.merge(df2, df1, left_on='card_nm', right_on='card_nm')
df_feb_conv.rename(columns={'use_x': 'issued', 'use_y': 'used'}, inplace=True)
df_feb_conv['issued_to_used'] = (df_feb_conv['used']/df_feb_conv['issued'])*100
df_feb_conv = df_feb_conv.round({'issued_to_used': 0})
df_feb_conv

issued  used  issued_to_used
card_nm                              
Alpha       732   166            23.0
Beta        240    20             8.0
Gamma       320   131            41.0

#### Март

In [11]:
df1 = df_mar.groupby('card_nm')['use'].sum().to_frame()
df2 = df_mar.groupby('card_nm')['use'].count().to_frame()
df_mar_conv = pd.merge(df2, df1, left_on='card_nm', right_on='card_nm')
df_mar_conv.rename(columns={'use_x': 'issued', 'use_y': 'used'}, inplace=True)
df_mar_conv['issued_to_used'] = (df_mar_conv['used']/df_mar_conv['issued'])*100
df_mar_conv = df_mar_conv.round({'issued_to_used': 0})
df_mar_conv

issued  used  issued_to_used
card_nm                              
Alpha       768   134            17.0
Beta        225    16             7.0
Gamma       383    37            10.0

#### Можно предположить, что программа по выдаче сертификатов была популярна у клиентов только до февраля. Конверсии за ноябрь, декабрь и январь выше, чем итоговая. Возникает вопрос: почему в феврале и марте процент использованных сертификатов упал? Для этого рассмотрим периоды, в течение которых выданные сертификаты используются.

In [12]:
print(df_nov[ (df_nov.use_dt >= '2018-11-01') & (df_nov.use_dt < '2018-12-01') ].shape[0], 'November cards were used in November')
print(df_nov[ (df_nov.use_dt >= '2018-12-01') & (df_nov.use_dt < '2019-01-01') ].shape[0], 'November cards were used in December')
print(df_nov[ (df_nov.use_dt >= '2019-01-01') & (df_nov.use_dt < '2019-02-01') ].shape[0], 'November cards were used in January')
print(df_nov[ (df_nov.use_dt >= '2019-02-01') & (df_nov.use_dt < '2019-03-01') ].shape[0], 'November cards were used in February')
print(df_nov[ (df_nov.use_dt >= '2019-03-01') & (df_nov.use_dt < '2019-04-01') ].shape[0], 'November cards were used in March')

93 November cards were used in November
194 November cards were used in December
43 November cards were used in January
0 November cards were used in February
0 November cards were used in March


In [13]:
print(df_dec[ (df_dec.use_dt >= '2018-12-01') & (df_dec.use_dt < '2019-01-01') ].shape[0], 'December cards were used in December')
print(df_dec[ (df_dec.use_dt >= '2019-01-01') & (df_dec.use_dt < '2019-02-01') ].shape[0], 'December cards were used in January')
print(df_dec[ (df_dec.use_dt >= '2019-02-01') & (df_dec.use_dt < '2019-03-01') ].shape[0], 'December cards were used in February')
print(df_dec[ (df_dec.use_dt >= '2019-03-01') & (df_dec.use_dt < '2019-04-01') ].shape[0], 'December cards were used in March')

105 December cards were used in December
199 December cards were used in January
38 December cards were used in February
0 December cards were used in March


#### Исходя из полученных данных можно предположить, что выданные сертификаты используются в течение 3 месяцев. Поэтому сертификаты, выданные в феврале и в марте, еще не все использовали. Поэтому посчитаем конверсию за ноябрь, декабрь и январь.

In [14]:
df_NOV_JAN = df[(df.receive_dt >= '2018-11-01') & (df.receive_dt < '2019-02-01')]
df1 = df_NOV_JAN.groupby('card_nm')['use'].sum().to_frame()
df2 = df_NOV_JAN.groupby('card_nm')['use'].count().to_frame()
df_NOV_JAN_conv = pd.merge(df2, df1, left_on='card_nm', right_on='card_nm')
df_NOV_JAN_conv.rename(columns={'use_x': 'issued', 'use_y': 'used'}, inplace=True)
df_NOV_JAN_conv['issued_to_used'] = (df_NOV_JAN_conv['used']/df_NOV_JAN_conv['issued'])*100
df_NOV_JAN_conv = df_NOV_JAN_conv.round({'issued_to_used': 0})
df_NOV_JAN_conv

issued  used  issued_to_used
card_nm                              
Alpha      1853   497            27.0
Beta        753    49             7.0
Gamma      1130   525            46.0

#### Исходная конверсия

In [15]:
df_conv

issued  used  issued_to_used
card_nm                              
Alpha      3570   812            23.0
Beta       1272    89             7.0
Gamma      1922   693            36.0

#### Получается, что сертификаты Gamma еще популярнее для клиентов, чем прежде -- 46% против 36%. Процент использвания сертификатов Alpha тоже повысился после пересчета.